File: 백준/Silver/15654. N과 M （5）/N과 M （5）.py
n, m = map(int, input().split())
arr = list(map(int, input().split()))
arr.sort()
visited = [0] * n
path = []

def abc(level):
    if level == m:
        print(' '.join(map(str, path)))
        return
    for i in range(n):
        if not visited[i]:
            visited[i] = 1
            path.append(arr[i])
            abc(level + 1)
            visited[i] = 0
            path.pop()

abc(0)

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static int[] arr;
    static int[] result;
    static boolean[] visited;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        result = new int[m];
        arr = new int[n];
        visited = new boolean[n];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i<n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        dfs(0);
        System.out.println(sb);
    }

    public static void dfs(int level) {
        if (level==m) {
            for (int i = 0; i<m; i++) {
                sb.append(result[i] + " ");
            }
            sb.append("\n");
            return;
        }
        for (int i = 0; i<n; i++) {
            if (!visited[i]) {
                result[level] = arr[i];
                visited[i] = true;
                dfs(level+1);
                visited[i] = false;
            }
        }
    }
}